### PR TITLE
fix coverage report for cmake / ninja combination

### PR DESCRIPTION
**Pull request by theKBro**
path to sourcefile is given relative from object dictionary

### CODE REVIEW

**Comment by theKBro** [comment]
+1 for someone who knows the repo and is willing to write tests for this :-)

**Comment by Spacetown** [comment]
We have a test with CMake. I don't understand your problem?

**Comment by theKBro** [comment]
I've added an addtional test, but unfortunately my development environment is also windows and thats why I cannot run the tests, the reference dir is copied from the other cmake test but maybe with the tests you understand why the changes are necessary

**Comment by theKBro** [comment]
Or could you support me on running the test with mingw for windows -> I got a problem on the tests `make clean` with the rm call

**Comment by theKBro** [comment]
@Spacetown now I really don't know how to proceed here anymore, could you please help out?

**Comment by Spacetown** [comment]
@theKBro I'll try to find a slot to look at this. Please be patient.

**Comment by Spacetown** [comment]
@theKBro Ive done several changes:
- Replace the addition of a key to options with the use of the original key in `process_gcov_data`.
- Install ninja in Docker container.
- Run ninja with option verbose to get info about command line and working directory.
- Add option `--object-directory=...` to gcovr call.

**Comment by Spacetown** [comment]
@theKBro Please can you rebase the branch?

File: gcovr/tests/cmake_oos_ninja/Makefile [hunk 1]
```diff
@@ -0,0 +1,33 @@
+
+all:
+	@echo "CC:   ${CC} (origin: $(origin CC))"
+	@echo "CXX:  ${CXX} (origin: $(origin CXX))"
+	@echo "GCOV: ${GCOV} (origin: $(origin GCOV))"
+	mkdir -p build/build
+	cd build/build && cmake -DCMAKE_TOOLCHAIN_FILE=../../Toolchain.cmake -G "Ninja" -S ../..
```
Comment (theKBro) [inline gcovr/tests/cmake_oos_ninja/Makefile:7]: @Spacetown why did you change this? any special reason?

Comment (Spacetown) [reply gcovr/tests/cmake_oos_ninja/Makefile:7]: Because in the other CMake test it is done the same way.

Comment (theKBro) [reply gcovr/tests/cmake_oos_ninja/Makefile:7]: ok, thats the "old" way

Comment (Spacetown) [reply gcovr/tests/cmake_oos_ninja/Makefile:7]: You can change it back., this was a try to fix the test. Pleas keep the direct ninja call to be sure that ninja is used.

Comment (theKBro) [reply gcovr/tests/cmake_oos_ninja/Makefile:7]: I'm fine with it as it is now.
fyi
The whole thing about the `cmake --build` call is that its using the correct buildtool defined by the generator.
Additional arguments can also be passed after another `--` so with appending `-- -v` you could also achieve verbose builds

Comment (Spacetown) [reply gcovr/tests/cmake_oos_ninja/Makefile:7]: > I'm fine with it as it is now.
> fyi
> The whole thing about the `cmake --build` call is that its using the correct buildtool defined by the generator.
> Additional arguments can also be passed after another `--` so with appending `-- -v` you could also achieve verbose builds

I was searching for something like this but if I run `cmake -h` there isn't a hint to the lonesome dash. For me it will be fine to use the new style in this test.

File: gcovr/gcov.py [hunk 2]
```diff
@@ -203,15 +204,21 @@ def guess_source_file_name_heuristics(
...
-    # 3. Try using the path to the gcda as the source directory
+    # 4. Try using the path to the gcda as the source directory
     fname = os.path.join(source_fname_dir, gcovname)
     if os.path.exists(fname):
         return os.path.normpath(fname)
 
-    # 4. Try using the path to the gcda file as the source directory, removing the path part from the gcov file
+    # 5. Try using the path to the gcda file as the source directory, removing the path part from the gcov file
     fname = os.path.join(source_fname_dir, os.path.basename(gcovname))
     return fname
```
Comment (Spacetown) [inline gcovr/gcov.py:223]: One thing I saw while fixing the test. If the file doesn't exist why is this path returned? I was wondering about the output path in the test result.
```suggestion
    # 5. Try using the path to the gcda file as the source directory, removing the path part from the gcov file
    fname = os.path.join(source_fname_dir, os.path.basename(gcovname))
    if os.path.exists(fname):
        return os.path.normpath(fname)

    # return the original path from the gcda file.
    return gcovname
```

File: gcovr/tests/cmake_oos_ninja/Makefile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	ninja -C build/build -v
```
Comment (Spacetown) [inline gcovr/tests/cmake_oos_ninja/Makefile:8]:
```suggestion
	cmake -DCMAKE_TOOLCHAIN_FILE=../../Toolchain.cmake -G "Ninja" -S . -B build/build
	cmake --build -- -v
```

**Review by Spacetown — CHANGES_REQUESTED** [review]
Please add yourself to the list of contributors and the issue to the changelog.
Why did you remove me as co author? 😢

File: gcovr/tests/cmake_oos_ninja/Makefile [hunk 4]
```diff
@@ -4,8 +4,8 @@ all:
 	@echo "CXX:  ${CXX} (origin: $(origin CXX))"
 	@echo "GCOV: ${GCOV} (origin: $(origin GCOV))"
 	mkdir -p build/build
-	cd build/build && cmake -DCMAKE_TOOLCHAIN_FILE=../../Toolchain.cmake -G "Ninja" -S ../..
-	ninja -C build/build -v
+	cmake -DCMAKE_TOOLCHAIN_FILE=../../Toolchain.cmake -G "Ninja" -S . -B build/build
+	cmake --build -- -v
```
Comment (Spacetown) [inline gcovr/tests/cmake_oos_ninja/Makefile:8]: My fault, the directory is missing.
```suggestion
	cmake --build build/build -- -v
```

**Comment by theKBro** [comment]
> Please add yourself to the list of contributors and the issue to the changelog.
> Why did you remove me as co author? 😢

sry im not used to git, as i'm working with svn...
how to readd you?

**Comment by Spacetown** [comment]
> > Please add yourself to the list of contributors and the issue to the changelog.
> > Why did you remove me as co author? 😢
> 
> sry im not used to git, as i'm working with svn...
> how to readd you?

You did it already by accepting my suggestion. This is done by adding the following line to the commit message:
```
Co-authored-by: Michael Förderer <40258682+Spacetown@users.noreply.github.com>
```

File: gcovr/gcov.py (hunk 2 above)
Comment (theKBro) [reply gcovr/gcov.py:223]: I also wondered about that part, but didn't want to change anything I could not test.
I am not sure if we should change that in this PR

Comment (theKBro) [reply gcovr/gcov.py:223]: but you are write -> if it does not exist there is no benefit

File: gcovr/gcov.py [hunk 5]
```diff
@@ -220,7 +220,11 @@ def guess_source_file_name_heuristics(
 
     # 5. Try using the path to the gcda file as the source directory, removing the path part from the gcov file
     fname = os.path.join(source_fname_dir, os.path.basename(gcovname))
-    return fname
+    if os.path.exists(fname):
+        return os.path.normpath(fname)
+
+    # return the original path from the gcda file.
+    return gcovname
```
[the remark this answers is not in the document]
Comment (Spacetown) [reply gcovr/gcov.py:227]: OK... Change it back and I'll do this in a separate PR.

[the remark this answers is not in the document]
Comment (theKBro) [reply gcovr/gcov.py]: done

**Review by Spacetown — APPROVED** [review]
